### DISCUSSION TITLE
0% Building static HTML for pages

### DISCUSSION THREAD

**Discussion by justinh00k** [gatsbyjs/gatsby#35921] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Had this problem with my project, so I created a new "npm init gatsby" project and it has the same issue. Running "gatsby build" hangs at "0% Building static HTML for pages" and never completes. Cloud builds through Gatsby and Cloudflare both seem to work, so I tried a few different Node versions locally but none fixed it. (Currently using LTS 16.15.1)

### Reproduction Link

https://github.com/justinh00k/gatsby-test-site

### Steps to Reproduce

1. npm init gatsby
2. gatsby build

### Expected Result

Project builds succesfully 

### Actual Result

success Writing page-data.json files to public directory - 0.046s - 3/3 64.55/s
Waiting for the debugger to disconnect...
[                            ]   0.001 s 0/3 0% Building static HTML for pages
Debugger listening on ws://127.0.0.1:49298/71fb6751-c5b3-44c7-a276-4267c2a49d95
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:49299/1576e931-61db-48ab-be55-291ed2fef95e
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:49300/0428a49a-6a71-426c-b6d8-15446315a06a
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:49301/aaf05f7c-48bc-4ea9-a179-8b1632bd24dc
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:49302/8ebe2df1-1fba-43d5-ac41-5fa0d942a394
For help, see: https://nodejs.org/en/docs/inspector

### Environment

```shell
System:
    OS: macOS 12.4
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
    Yarn: 1.22.17 - ~/.volta/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
  Browsers:
    Chrome: 102.0.5005.61
    Firefox: 101.0.1
    Safari: 15.5
  npmPackages:
    gatsby: ^4.16.0 => 4.16.0
  npmGlobalPackages:
    gatsby-cli: 4.16.0
```

**Comment by LekoArts** [comment, marked as answer]
Your build log shows your problem:

```shell
Debugger listening on ws://127.0.0.1:49298/71fb6751-c5b3-44c7-a276-4267c2a49d95
For help, see: [nodejs.org/en/docs/inspector](https://nodejs.org/en/docs/inspector)
```

You're attaching a debugger to your run so you'll need to change your configuration. If you run it via VSCode, try it just in the terminal.

**Reply by justinh00k** [reply]
TYSM. I didn't recognize that debugger stuff, and had never configured it, but couldn't turn it off through the VSCode menu. Turns out, the words "Auto Attach" were appearing in the VS Code status bar (very bottom of the window.) Clicking that let me turn that off, which solved the problem.

**Comment by luciobenini** [comment]
I have the same problem. Build stops at 0% and no progress

**Reply by justinh00k** [reply]
I actually started having this problem again also - not the debugger this time.

**Reply by justinh00k** [reply]
I downgraded to gatsby 4.10.0 and React 17.0.2. This time instead of hanging, I got an error (about localStorage not being available in SSR). Fixed that and it built properly. Then I was able to upgrade to gatsby 4.17.2 and react 18.2.0 and build successfully.

**Reply by diogomoreira** [reply]
Same here. Tried the approach suggested by @justinh00k but without success.

**Reply by nrandell** [reply]
I'm seeing similar things - if there is an error thrown building a page, the entire thing just hangs. I tried a simple reproduction and of course that works ok. I'm trying to find a minimal reproduction

**Reply by kevinrkowalski** [reply]
I'm having the same issue, would love to hear any progress!

**Reply by nrandell** [reply]
I've just reproduced and raised an issue https://github.com/gatsbyjs/gatsby/issues/36075

It seems to be a page added via createPage causes the issue. Hopefully I describe well enough in the issue to understand what is going on!

**Reply by JohannBoumans** [reply]
Same problem for me, i commented all pages in gatsby node then build succeed, but problem seems to be related to createpage made two weeks ago with no changes until now and builds didn't failed until yesterday thats weird

**Reply by luciobenini** [reply]
I confirm. When an error on pages occurs the build fails (even if the error is in plugins).

**Reply by letzdoo-js** [reply]
Same here. Tried the approach suggested by @justinh00k and with success, the error is reported.

**Reply by achoshengelia** [reply]
I had the same problem, tried downgrading to React 17 and ran the build command again. It finally logged the bugs that were halting the build process and needed to be fixed. Seems like there's a problem with logging build errors with React 18.

**Comment by ms31-dev** [comment]
I have the same problem, that every time I run my "**gastby build**" command (**I ran before "gastby clean"**), that it stucks on the task "Building static HTML for pages" with 0 percent and does not finish the build.

My stack is:

"gatsby": "^4.18.0"
"react": "^18.1.0"
"react-dom": "^18.1.0"
"theme-ui": "^0.3.4"
"@emotion/core": "^10.3.1",
"@emotion/react": "^11.9.3",


I identified that the problem is caused by the following code in my **Header.Menu.jsx**:

```
const NaviList = ({ navKey, wrapperProps, items, items2, ...props }) =>
  items ? (
    <Flex {...wrapperProps}>
      {items.map((menuItem, index) => (
        <>
          <NaviItem key={`${navKey}-${index}`} {...menuItem} {...props} />
          <Box sx={styles.secondUl}>
            {items2.map((menuItem2, index2) => (
              <NaviItem key={`${navKey}-${index2}`} {...menuItem2} {...props} />
            ))}
          </Box>
        </>
      ))}
    </Flex>
  ) : null
```

When I uncomment the second JavaScript map `items2.map()` then my gastby build runs without problems and finishes the build compeletly without getting stucked.

```
const NaviList = ({ navKey, wrapperProps, items, items2, ...props }) =>
  items ? (
    <Flex {...wrapperProps}>
      {items.map((menuItem, index) => (
        <>
          <NaviItem key={`${navKey}-${index}`} {...menuItem} {...props} />
          <Box sx={styles.secondUl}>
            {/*{items2.map((menuItem2, index2) => (*/}
            {/*  <NaviItem key={`${navKey}-${index2}`} {...menuItem2} {...props} />*/}
            {/*))}*/}
          </Box>
        </>
      ))}
    </Flex>
  ) : null
```

But how can I achive that my JSX component can use these nested JavaScript Maps `items.map()` and `items2.map()` function?

**Comment by yutoji** [comment]
I fixed it by avoiding use of `window` object in pages. Then `yarn build` succeeded.
Instead of using window directly,  I check if the window is defined.

```JavaScript
// util.ts
const isBrowser = typeof window !== "undefined";
export const getWindow = <T>(
  callback: (window: Window) => T
): T | undefined => {
  if (!isBrowser) return undefined;
  return callback(window);
};

// index.tsx
const IndexPage () => {
  const windowLocationSearch = getWindow((window) => window.location.search) || ""
  // const windowLocationSearch = window.location.search // <= this code dosn't work (yarn build fails)
  ...
}
```

**Reply by jordannpenn** [reply]
I had the same issue and this fixed it! All I did was check if window is defined and then the build worked!
